test/gleam/dict_test.gleam

```diff
@@ -199,6 +199,105 @@ pub fn upsert_test() {
   |> should.equal(dict.from_list([#("a", 0), #("b", 1), #("c", 2), #("z", 0)]))
 }
 
+pub fn update_test() {
+  let dict = dict.from_list([#("a", 0), #("b", 1), #("c", 2)])
+
+  let inc_if_exists_or_set = fn(x) {
+    case x {
+      Some(i) -> Ok(i + 1)
+      None -> Ok(1)
+    }
+  }
+
+  dict
+  |> dict.update("a", inc_if_exists_or_set)
+  |> should.equal(dict.from_list([#("a", 1), #("b", 1), #("c", 2)]))
+
+  dict
+  |> dict.update("e", inc_if_exists_or_set)
+  |> should.equal(dict.from_list([#("a", 0), #("b", 1), #("c", 2), #("e", 1)]))
+
+  let discard_if_exists_else_noop = fn(x) {
+    case x {
+      Some(_i) -> Error(Nil)
+      None -> Error(Nil)
+    }
+  }
+
+  dict
+  |> dict.update("a", discard_if_exists_else_noop)
+  |> should.equal(dict.from_list([#("b", 1), #("c", 2)]))
+
+  dict
+  |> dict.update("e", discard_if_exists_else_noop)
+  |> should.equal(dict.from_list([#("a", 0), #("b", 1), #("c", 2)]))
+
+  let inc_if_exists_else_noop = fn(x) {
+    case x {
+      Some(i) -> Ok(i + 1)
+      None -> Error(Nil)
+    }
+  }
+
+  dict
+  |> dict.update("a", inc_if_exists_else_noop)
+  |> should.equal(dict.from_list([#("a", 1), #("b", 1), #("c", 2)]))
+
+  dict
+  |> dict.update("e", inc_if_exists_else_noop)
+  |> should.equal(dict.from_list([#("a", 0), #("b", 1), #("c", 2)]))
+
+  let discard_if_exists_else_set_to_zero = fn(x) {
+    case x {
+      Some(_i) -> Error(Nil)
+      None -> Ok(0)
+    }
+  }
+
+  dict
+  |> dict.update("a", discard_if_exists_else_set_to_zero)
+  |> should.equal(dict.from_list([#("b", 1), #("c", 2)]))
+
+  dict
+  |> dict.update("e", discard_if_exists_else_set_to_zero)
+  |> should.equal(dict.from_list([#("a", 0), #("b", 1), #("c", 2), #("e", 0)]))
+
+  let discard_if_exists_else_set_to_zero = fn(x) {
+    case x {
+      Some(_i) -> Error(Nil)
+      None -> Ok(0)
+    }
+  }
+
+  dict
+  |> dict.update("a", discard_if_exists_else_set_to_zero)
+  |> should.equal(dict.from_list([#("b", 1), #("c", 2)]))
+
+  dict
+  |> dict.update("e", discard_if_exists_else_set_to_zero)
+  |> should.equal(dict.from_list([#("a", 0), #("b", 1), #("c", 2), #("e", 0)]))
+
+  // Utilize a function returning a `Result` in the change function.
+  let half_or_zero_else_noop = fn(x) {
+    case x {
+      Some(i) -> int.divide(i, 2)
```

Comment:
Because other functions usually return `Result` rather than `Option`, I have opted for the change function to return `Result` not `Option` as well.

I have added a unit test example to show what I mean.

Are you certain you want to return `Option` instead?

ref https://github.com/gleam-lang/stdlib/pull/659#discussion_r1691409286 cc @lpil